src/systems/a320_systems/src/flaps_computer.rs

```diff
@@ -0,0 +1,1327 @@
+            speed,
+            max_angle,
+            left_position_percent_id: format!("LEFT_{}_POSITION_PERCENT", surface_type),
+            right_position_percent_id: format!("RIGHT_{}_POSITION_PERCENT", surface_type),
+            left_position_angle_id: format!("LEFT_{}_ANGLE", surface_type),
+            right_position_angle_id: format!("RIGHT_{}_ANGLE", surface_type),
+
+            hyd_green_pressure: 0.,
+        }
+    }
+
+    fn update(&mut self, context: &UpdateContext, sfcc_demand: Option<Angle>) {
+        if self.hyd_green_pressure > 2000. {
+            if let Some(demanded_angle) = sfcc_demand {
+                let actual_minus_target: Angle = demanded_angle - self.current_angle;
+                if actual_minus_target.get::<degree>().abs() > Self::ANGLE_DELTA_DEGREE {
+                    self.current_angle += Angle::new::<degree>(
+                        actual_minus_target.get::<degree>().signum()
+                            * self.speed
+                            * context.delta_as_secs_f64(),
+                    );
+                    if self.current_angle < Angle::new::<degree>(0.) {
+                        self.current_angle = Angle::new::<degree>(0.);
+                    }
+                } else {
+                    self.current_angle = demanded_angle;
+                }
+            }
+        }
+    }
+}
+
+impl SimulationElement for SlatFlapGear {
+    fn write(&self, writer: &mut SimulatorWriter) {
+        writer.write(
+            &self.left_position_percent_id,
+            self.current_angle / self.max_angle,
+        );
+        writer.write(
+            &self.right_position_percent_id,
+            self.current_angle / self.max_angle,
+        );
+        writer.write(&self.left_position_angle_id, self.current_angle);
+        writer.write(&self.right_position_angle_id, self.current_angle);
+    }
+
+    fn read(&mut self, reader: &mut SimulatorReader) {
+        self.hyd_green_pressure = reader.read("HYD_GREEN_PRESSURE");
+    }
+}
+
+pub struct SlatFlapComplex {
+    sfcc: SlatFlapControlComputer,
+    flaps_handle: FlapsHandle,
+    flap_gear: SlatFlapGear,
+    slat_gear: SlatFlapGear,
+}
+
+impl SlatFlapComplex {
+    pub fn new() -> Self {
+        Self {
+            sfcc: SlatFlapControlComputer::new(),
+            flaps_handle: FlapsHandle::new(),
+            flap_gear: SlatFlapGear::new(2., Angle::new::<degree>(40.), "FLAPS"),
+            slat_gear: SlatFlapGear::new(1.5, Angle::new::<degree>(27.), "SLATS"),
+        }
+    }
+
+    pub fn update(&mut self, context: &UpdateContext) {
+        self.sfcc
+            .update(context, self.flaps_handle.signal_new_position());
+        self.flap_gear.update(
+            context,
+            self.sfcc.signal_flap_movement(self.flap_gear.current_angle),
+        );
+        self.slat_gear.update(
+            context,
+            self.sfcc.signal_slat_movement(self.slat_gear.current_angle),
+        );
+    }
```

Comment:
I am pretty new to Rust so I may asking dumb questions.
Why would using a trait with `current_angle()` be needed if it is only going to be implemented by `SlatFlapGear`?
Why wouldn't `sfcc.update` be accepting two `&SlatFlapGear` instead of `&impl X`?
Any obvious benefits for this methodology?

Reply:
There's no problem in asking questions. Please feel free to do so. Asking helps all of us learn!

You are right, passing `&SlatFlapGear` is also possible. From a Rust perspective there's no specific need to create a trait and pass it. I had the following considerations while making my suggestions:

- When using the current implementation, some overarching type (in this case the `SlatFlapComplex`) becomes responsible for passing the right information to the `sfcc`, `flap_gear`, and `slat_gear`. It needs to know about the internal functioning of the `SlatFlapControlComputer` and the `SlatFlapGear`. If `SlatFlapComplex` changes, or is removed, or its content is split across multiple types, then the way of passing information needs to be modified (for example, you might need to start passing two `Angle`s around through multiple layers of `update`).
- There is nothing guarding you against passing the wrong information.
- Simply passing `&SlatFlapGear` to `SlatFlapControlComputer`, and `&SlatFlapControlComputer` to `SlatFlapGear` creates a bidirectional dependency between `SlatFlapGear` and `SlatFlapControlComputer`, which is something I usually try to avoid.
- Passing concrete types exposes more of the interface than needed for function to execute, i.e. `SlatFlapControlComputer` shouldn't be able to access `&SlatFlapGear.update`. By using traits one only exposes what one wants the client to use. This is an application of the [Interface Segregation Principle](https://en.wikipedia.org/wiki/Interface_segregation_principle).

Reply:
I made some changes to this in  `28a36b9`. This is probably not entirely what you suggested, but it is close. Since `SlatFlapGear` is just a placeholder, and will go away when integrating with the hydraulic model, I think we shouldn't spend too much time on this.